src/health.py

```diff
@@ -0,0 +1,170 @@
+errors_we_dont_want_in_reasons = ['FailedScheduling', 'SchedulerError', 'CrashLoopBackOff', 'FailedCreate', 'OOMKilled']
+
+def check_until_job_is_operational(
+    job: JobDto,
+    deployment_timestamp: int = 0,
+    on_job_alive: Callable = None,
+    headers: dict[str, str] | None = None,
+):
+    """
+    Mostly copied from racetrack/lifecycle/lifecycle/monitor/health.py, but we wanted to checks for kubernetes errors in the liviness check
+    """
+    resource_name = job_resource_name(job.name, job.version)
+    base_url: str = f'http://{job.internal_name}'
+    k8s_client = k8s_api_client()
+    core_api = client.CoreV1Api(k8s_client)
+
+    # this can have long timeout since any potential malfunction in here is not related to a model/entrypoint
+    # but the cluster errors that shouldn't happen usually
+    @backoff.on_exception(backoff.fibo, RuntimeError, max_value=3,
+                          max_time=LifecycleCache.config.timeout_until_job_alive, jitter=None, logger=None)
+    def _wait_until_job_is_alive(_base_url: str, core_api, resource_name, expected_deployment_timestamp: int, _headers: dict[str, str] | None) -> Response:
+        return check_job_is_alive(_base_url, core_api, resource_name, expected_deployment_timestamp, _headers)
+
+    response = _wait_until_job_is_alive(base_url, core_api, resource_name, deployment_timestamp, headers)
+    _validate_live_response(response)
+    if on_job_alive is not None:
+        on_job_alive()
+
+    @backoff.on_exception(backoff.fibo, TimeoutError, max_value=3,
+                          max_time=LifecycleCache.config.timeout_until_job_ready, jitter=None, logger=None)
+    def _wait_until_job_is_ready(_base_url: str, _headers: dict[str, str] | None = None):
+        check_job_is_ready(_base_url, _headers)
+
+    _wait_until_job_is_ready(base_url, headers)
+
+
+def check_job_is_alive(
+    base_url: str,
+    core_api,
+    resource_name,
+    expected_deployment_timestamp: int,
+    headers: dict[str, str] | None,
+) -> Response:
+
+
+    """Check the containers in the pods of the job"""
+    pods_by_job = get_job_pods(core_api)
+    pods = pods_by_job.get(resource_name, [])
+
+    if not isinstance(pods, list):
+        pods = []
+
+    for pod in pods:
+        if pod.status:
+            container_statuses = getattr(pod.status, 'container_statuses', [])
+            if not isinstance(container_statuses, list):
+                container_statuses = []
+
+            for container_status in container_statuses:
+                # Check both 'state' and 'last_state' for terminated and waiting
+                check_state(getattr(container_status.state, 'terminated', None))
+                check_state(getattr(container_status.state, 'waiting', None))
+                check_state(getattr(container_status.last_state, 'terminated', None))
+                check_state(getattr(container_status.last_state, 'waiting', None))
```

Comment:
This is new

Reply:
Most of these attributes are optional, https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1ContainerStatus.md, therefore the getattrs.

Reply:
I assume by optional they mean it can be None, but this is something different than the missing attribute. I'm leaning into accessing the attributes strictly, with type annotations if possible. For instance:
`state: V1ContainerState | None = container.state`
This will prevent from potential typos in the attribute names